### ISSUE TITLE
`reportUnusedCoroutine` has false negative around `bool(coroutine)`.

### ISSUE THREAD

**Issue by patrick-kidger** [microsoft/pyright#9579]
**Code or Screenshots**

```python
async def f():
    return 1

if f():
    pass
```
does not trigger `reportUnusedCoroutine`. Instead the coroutine is silently dropped on the floor and is never awaited.

Expected behaviour is for this to raise an error. From the docs: 'reportUnusedCoroutine: Generate or suppress diagnostics for call statements whose return value is not used in any way and is a Coroutine.'

**Alternatives considered**

I did find that this is checked under `reportUnnecessaryComparison`:

https://github.com/microsoft/pyright/commit/9314f14b71cbd70ddf43ca2d5ed8ec4d0528f1b9

but at least for me I think that is not desirable, since it enforces implicit-cast-to-bool. The possibility of that is precisely what introduced the bug in the above code snippet, so I'm not looking to encourage it further!

**VS Code extension or command-line**

Via pre-commit, version 1.1.390.

**Comment by erictraut** [comment]
I don't consider this a bug. This check is working as intended. The coroutine isn't dropped on the floor in this case. It is passed as an argument to the `bool` method, and the return result of that is used in a conditional check. I can see why this could lead to a bug, but this condition is not covered by the `reportUnusedCoroutine` check.

I could maybe see a new check being added specifically for this case, so I'll convert this to an enhancement request.

I might consider this enhancement request in the future if there's sufficient demand for such a check from other pyright users (in the form of upvotes to this issue). For now, I'm going to reject it because I'm not convinced this is a sufficiently-common source of bugs to justify a new check.

**Comment by patrick-kidger** [comment]
That's a shame. The Python runtime considers this a bug, at least:
```python
.../file.py:4: RuntimeWarning: coroutine 'f' was never awaited
  if f():
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

**Comment by fabxc** [comment]
As another more data point: especially when migrating entire code bases to async, this can extremely easily slip through (and did for us in several instances). But also afterwards forgetting an `await` is very common. This was a significant factor in abandoning our migration to async as we believe it will make our code base overall more error prone going forward. It would be great if pyright could help out on this in the future.

**Comment by erictraut** [comment]
@fabxc, the `reportUnnecessaryComparison` check can be used to detect cases like this. It reports situations where a condition used in an `if` statement always evaluates to true or false. This includes cases where a coroutine is called but is not awaited.
